libtenzir/builtins/operators/chart2.cpp

```diff
@@ -0,0 +1,893 @@
+                        enumeration_type>()) {
+        diagnostic::warning("cannot group type `{}`", gs.type.kind())
+          .primary(args_.group.value())
+          .emit(dh);
+        check(b->AppendNulls(gs.length()));
+        continue;
+      }
+      for (auto&& value : gs.values()) {
+        if (is<caf::none_t>(value)) {
+          check(b->AppendNull());
+          continue;
+        }
+        const auto f = detail::overload{
+          [&](const enumeration& x) {
+            return std::string{as<enumeration_type>(gs.type).field(x)};
+          },
+          [&](const int64_t& x) {
+            return fmt::to_string(x);
+          },
+          [&](const auto&) {
+            return fmt::to_string(value);
+          },
+        };
+        check(b->Append(match(value, f)));
+      }
+    }
+    return series{string_type{}, finish(*b)};
+  }
+
+  auto get_groups(group_map& map, const data_view x, session ctx) const
+    -> grouped_bucket* {
+    // PERF: Maybe we only need to materialize when inserting new
+    const auto xv = materialize(x);
+    if (auto it = map.find(xv); it != map.end()) {
+      return &it->second;
+    }
+    if (map.size() == args_.limit.inner) {
+      diagnostic::warning("got more than {} data points", args_.limit.inner)
+        .primary(args_.x)
+        .note("skipping excess data points")
+        .hint(
+          "consider filtering data or aggregating over a bigger `resolution`")
+        .emit(ctx);
+    }
+    return &map[xv];
+  }
+
+  auto get_bucket(group_map& map, const plugins_map& plugins, const data_view x,
+                  const std::string_view group, session ctx) const -> bucket* {
+    if (args_.ty != chart_type::bar and args_.ty != chart_type::pie) {
+      if (is<caf::none_t>(x)) {
+        diagnostic::warning("x-axis cannot be `null`")
+          .primary(args_.x)
+          .emit(ctx);
+        return nullptr;
+      }
+    }
+    auto gs = get_groups(map, x, ctx);
+    if (not gs) {
+      return nullptr;
+    }
+    if (auto it = gs->find(group); it != gs->end()) {
+      return &it->second;
+    }
+    auto [it, _] = gs->emplace(group, args_.make_bucket(plugins, ctx));
+    return &it->second;
+  }
+
+  auto filter_input(generator<table_slice> input, diagnostic_handler& dh) const
+    -> generator<table_slice> {
+    const auto expr = std::invoke([&]() -> ast::expression {
+      if (args_.x_min and args_.x_max) {
+        return ast::binary_expr{
+          args_.x_min->expr,
+          {ast::binary_op::and_, location::unknown},
+          args_.x_max->expr,
+        };
+      }
+      if (args_.x_min) {
+        return args_.x_min->expr;
+      }
+      TENZIR_ASSERT(args_.x_max);
+      return args_.x_max->expr;
+    });
+    for (auto&& slice : input) {
+      if (slice.rows() == 0) {
+        co_yield {};
+        continue;
+      }
+      if (auto filtered = filter(slice, args_.filter)) {
+        slice = std::move(*filtered);
+      } else {
```

Comment:
Isn't the filter relative to the resulting event? You construct it from a conjunction of the predicates after this operator, so it should be applied to the output.

Reply:
True, now filtering the output slices.